### ISSUE TITLE
string format in mask prediction task using CodeBERT (MLM)

### ISSUE THREAD

**Issue by pelican9** [microsoft/CodeBERT#84]
Hi,

I want to mask tokens in a multiple-line function and it would be great if I can confirm that I don't need to add any special token like \<s> or \</s> anywhere to the string of code. I am aware that it's been mentioned in the paper that 'We regard a piece of code as a sequence of tokens', but I am not sure if any special tokens e.g. \</s> needs to be added to separate lines.

For example, if the data is

```
{'code': 'def get_vid_from_url(url):\n'
         '        """Extracts video ID from URL.\n'
         '        """\n'
         "        return match1(url, r'youtu\\.be/([^?/]+)') or \\\n"
         "          match1(url, r'youtube\\.com/embed/([^/?]+)') or \\\n"
         "          match1(url, r'youtube\\.com/v/([^/?]+)') or \\\n"
         "          match1(url, r'youtube\\.com/watch/([^/?]+)') or \\\n"
         "          parse_query_param(url, 'v') or \\\n"
         "          parse_query_param(parse_query_param(url, 'u'), 'v')",
 'code_tokens': ['def',
                 'get_vid_from_url',
                 '(',
                 'url',
                 ')',
                 ':',
                 'return',
                 'match1',
                 '(',
                 'url',
                 ',',
                 "r'youtu\\.be/([^?/]+)'",
                 ')',
                 'or',
                 'match1',
                 '(',
                 'url',
                 ',',
                 "r'youtube\\.com/embed/([^/?]+)'",
                 ')',
                 'or',
                 'match1',
                 '(',
                 'url',
                 ',',
                 "r'youtube\\.com/v/([^/?]+)'",
                 ')',
                 'or',
                 'match1',
                 '(',
                 'url',
                 ',',
                 "r'youtube\\.com/watch/([^/?]+)'",
                 ')',
                 'or',
                 'parse_query_param',
                 '(',
                 'url',
                 ',',
                 "'v'",
                 ')',
                 'or',
                 'parse_query_param',
                 '(',
                 'parse_query_param',
                 '(',
                 'url',
                 ',',
                 "'u'",
                 ')',
                 ',',
                 "'v'",
                 ')'],

```

My understanding is that firstly preprocess code in the dataset using this 
https://github.com/microsoft/CodeBERT/blob/e2348efb19726b763f296cd94510e222d433cb90/CodeBERT/code2nl/run.py#L69-L70

Then replace a token e.g. 'return' with '\<mask\>'. And the script for getting the alternatives' probability distribution for the masked token 'return' is simply 

```
from transformers import RobertaTokenizer, RobertaForMaskedLM, pipeline

model = RobertaForMaskedLM.from_pretrained('microsoft/codebert-base-mlm')
tokenizer = RobertaTokenizer.from_pretrained('microsoft/codebert-base-mlm')

code = "def get_vid_from_url ( URL ) : <mask> match1 ( url , r'youtu\.be/([^?/]+)' ) or match1 ( url , r'youtube\.com/embed/([^/?]+)' ) or match1 ( url , r'youtube\.com/v/([^/?]+)' ) or match1 ( url , r'youtube\.com/watch/([^/?]+)' ) or parse_query_param ( url , 'v' ) or parse_query_param ( parse_query_param ( url , 'u' ) , 'v' )"
fill_mask = pipeline('fill-mask', model=model, tokenizer=tokenizer)

outputs = fill_mask(code)

```

Thank you in advance!

**Comment by pelican9** [comment]
And there's no need to add a token to indicate indentation? I'm a bit curious about this because I thought a token indicating an indentation or a new row could add more information to the dataset?

**Comment by guoday** [comment]
Because we remove all "\n" and "\t" from the pre-training data (i.e. `" ".join('code_tokens')`), you don't need to add any special token like <s> or </s>.

**Comment by guoday** [comment]
> And there's no need to add a token to indicate indentation? I'm a bit curious about this because I thought a token indicating an indentation or a new row could add more information to the dataset?


You don't need to   add a token to indicate indentation. Yes, indicating an indentation or a new row could add more information, but we need to pre-process the data before pre-training. However, in our model, we just simply use `" ".join('code_tokens')` as the input.

**Comment by pelican9** [comment]
Hi thank you for your quick response! 

I want to mask each token in a given code and find the probability of the token appearing in the given code one by one. But I don't know how to preprocess the code correctly.

For example, the given code is
```
def a_function(a,b):
    if a == b:
        return a 
    elif a<b:
        return b 
    else: 
        return a+b
```
Because MLM uses the Roberta tokenizer, which sometimes splits PL tokens (e.g. "elif") into multiple tokens (e.g. "el" and "if"), I need to first tokenize the code using the Roberta tokenizer and then mask out the tokens one by one. I find that how I preprocess the code can make a big difference to the result:

1. First method. Preprocess the code by splitting symbols e.g.":" so that the code shares the same format as the code that is fed into the code2nl model:
https://github.com/microsoft/CodeBERT/blob/e2348efb19726b763f296cd94510e222d433cb90/CodeBERT/code2nl/run.py#L69-L70

```
CODE = """def a_function(a,b):
    if a == b:
        return a 
    elif a<b:
        return b 
    else: 
        return a+b"""

# Preprocess step that generates code with the same format as the code that is fed into the code2nl model
pat = re.compile(r'''(==|([[\]{\}=.():&"/%*,<>!+-]))''') 
CODE = ' '.join(pat.sub(" \\1 ", CODE.lstrip()).split())
CODE = ' '.join(CODE.strip().split()) 
print(f"CODE\n{CODE}\n")

# Robertatokenizer
masked_token_list = [x.replace("Ġ"," ") for x in tokenizer.tokenize(CODE)]
print(f"masked_token_list\n{masked_token_list}\n")

for i,x in enumerate(masked_token_list):
    masked_code_string = "".join(masked_token_list[:i])+" <mask> "+"".join(masked_token_list[i+1:])
    print(f"masked_code_string\n{masked_code_string}\n")
    outputs_with_target = fill_mask(masked_code_string,targets=x)
    print(f"outputs_with_target\n{outputs_with_target}\n")
    
    output_top5 = fill_mask(masked_code_string)
    for i in range(5):
        print(f"outputs top {i}\n{output_top5[i]}\n")

    break
```

Output:
```
CODE
def a_function ( a , b ) : if a == b : return a elif a < b : return b else : return a + b

masked_token_list
['def', ' a', '_', 'function', ' (', ' a', ' ,', ' b', ' )', ' :', ' if', ' a', ' ==', ' b', ' :', ' return', ' a', ' el', 'if', ' a', ' <', ' b', ' :', ' return', ' b', ' else', ' :', ' return', ' a', ' +', ' b']

masked_code_string
 <mask>  a_function ( a , b ) : if a == b : return a elif a < b : return b else : return a + b

outputs_with_target
[{'score': 6.355445657391101e-05, 'token': 9232, 'token_str': 'def', 'sequence': 'defa_function ( a, b ) : if a == b : return a elif a < b : return b else : return a + b'}]

outputs top 0
{'score': 0.36111196875572205, 'token': 1215, 'token_str': '_', 'sequence': '_a_function ( a, b ) : if a == b : return a elif a < b : return b else : return a + b'}

outputs top 1
{'score': 0.027643155306577682, 'token': 176, 'token_str': '2', 'sequence': '2a_function ( a, b ) : if a == b : return a elif a < b : return b else : return a + b'}

outputs top 2
{'score': 0.02554234303534031, 'token': 3697, 'token_str': 'vers', 'sequence': 'versa_function ( a, b ) : if a == b : return a elif a < b : return b else : return a + b'}

outputs top 3
{'score': 0.018836617469787598, 'token': 5489, 'token_str': 'par', 'sequence': 'para_function ( a, b ) : if a == b : return a elif a < b : return b else : return a + b'}

outputs top 4
{'score': 0.018411273136734962, 'token': 3894, 'token_str': 'ct', 'sequence': 'cta_function ( a, b ) : if a == b : return a elif a < b : return b else : return a + b'}
```
2. Second method. Do not split symbols. This is the same string format as the MLM probing example. The output probability for the masked token "def" is higher than it from the first method.
```
CODE = """def a_function(a,b):
    if a == b:
        return a 
    elif a<b:
        return b 
    else: 
        return a+b"""

# DO NOT split PL tokens
# pat = re.compile(r'''(==|([[\]{\}=.():&"/%*,<>!+-]))''') 
# CODE = ' '.join(pat.sub(" \\1 ", CODE.lstrip()).split())
CODE = ' '.join(CODE.strip().split()) 
print(f"CODE\n{CODE}\n")

# Robertatokenizer
masked_token_list = [x.replace("Ġ"," ") for x in tokenizer.tokenize(CODE)]
print(f"masked_token_list\n{masked_token_list}\n")

for i,x in enumerate(masked_token_list):
    masked_code_string = "".join(masked_token_list[:i])+" <mask> "+"".join(masked_token_list[i+1:])
    print(f"masked_code_string\n{masked_code_string}\n")
    outputs_with_target = fill_mask(masked_code_string,targets=x)
    print(f"outputs_with_target\n{outputs_with_target}\n")
    
    output_top5 = fill_mask(masked_code_string)
    for i in range(5):
        print(f"outputs top {i}\n{output_top5[i]}\n")

    break
```
Output:
```
CODE
def a_function(a,b): if a == b: return a elif a<b: return b else: return a+b

masked_token_list
['def', ' a', '_', 'function', '(', 'a', ',', 'b', '):', ' if', ' a', ' ==', ' b', ':', ' return', ' a', ' el', 'if', ' a', '<', 'b', ':', ' return', ' b', ' else', ':', ' return', ' a', '+', 'b']

masked_code_string
 <mask>  a_function(a,b): if a == b: return a elif a<b: return b else: return a+b

outputs_with_target
[{'score': 0.0010072329314425588, 'token': 9232, 'token_str': 'def', 'sequence': 'defa_function(a,b): if a == b: return a elif a<b: return b else: return a+b'}]

outputs top 0
{'score': 0.40951812267303467, 'token': 1215, 'token_str': '_', 'sequence': '_a_function(a,b): if a == b: return a elif a<b: return b else: return a+b'}

outputs top 1
{'score': 0.042673781514167786, 'token': 16963, 'token_str': 'fast', 'sequence': 'fasta_function(a,b): if a == b: return a elif a<b: return b else: return a+b'}

outputs top 2
{'score': 0.025901032611727715, 'token': 31362, 'token_str': 'circ', 'sequence': 'circa_function(a,b): if a == b: return a elif a<b: return b else: return a+b'}

outputs top 3
{'score': 0.01927659660577774, 'token': 5489, 'token_str': 'par', 'sequence': 'para_function(a,b): if a == b: return a elif a<b: return b else: return a+b'}

outputs top 4
{'score': 0.015669340267777443, 'token': 8660, 'token_str': 'iter', 'sequence': 'itera_function(a,b): if a == b: return a elif a<b: return b else: return a+b'}
```
So
1. Which or any other method should I use to preprocess the code for my use case? 
2. Why do you think the probability of predicting a token "def" with either method is so low? 
3. Is it possible to mask whitespace-separated PL tokens (including tokens that will be decomposed by the Roberta tokenizer) (e.g. "elif") and get the probability of this token (e.g. "elif")?

Any advice would be appreciated! Thank you in advance!
